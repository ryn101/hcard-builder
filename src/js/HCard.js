/**
 * Container for HCard specific information
 */
class HCard {
    constructor(
        givenName,
        surname,
        email,
        phone,
        houseNumber,
        street,
        suburb,
        state,
        postcode,
        country,
    ) {
        this.personalDetails = {
            givenName,
            surname,
            email,
            phone,
        };

        this.address = {
            houseNumber,
            street,
            suburb,
            state,
            postcode,
            country,
        };

        Object.defineProperty(this, 'fullName', {
            enumerable: false,
            get: () => [this.personalDetails.givenName, this.personalDetails.surname].filter(Boolean).join(' '),
        });
    }

    /**
     * Returns the full name being a combination of given and surnames
     *
     * @returns {string}
     */
    fullName = () => {
        const { givenName, surname } = this.personalDetails;
        return [givenName, surname].filter(Boolean).join(' ');
    }
}

export default HCard;
