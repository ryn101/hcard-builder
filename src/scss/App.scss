@import 'colours';

@mixin single-row-form-fields() {
  .hcard-builder {
    .form__field, .form__button-wrapper {
        flex: 1 100% !important;
        margin-right: 0 !important;
    }

    .form__button-wrapper:last-of-type {
      margin-top: 10px;
    }
  }
}

html, body, main {
  min-height: 100vh;
}

body {
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  :focus {
    &, & + * {
      outline: rgb(112, 171, 245) auto 5px;
    }
  }
}

main {
  background-color: rgb(225,227,232);
  display: flex;
  font-family: MerriweatherSans, Arial, sans-serif;

  input:not([type=file]) {
    border: 1px solid $boxed-out-grey;
    border-radius: 3px;
    padding: 3px 8px;
    font-family: Merriweather, Arial, serif;
    -webkit-appearance: none;
  }

  .hcard-preview {
    flex: 1;
    flex-basis: 52%;
    justify-content: center;
    flex-direction: column;
    min-width: 502px;
    padding: 145px 10px 0 10px;

    .vcard {
      margin: 0 auto;
    }
  }
}

@media screen and (max-width: 1000px) and (min-width: 751px) {
  main {
    @include single-row-form-fields;

    .hcard-preview {
      padding-top: 0;
    }

    .vcard {
      position: fixed;
      // Height of the vcard element (306px) divided by 2
      // Position above the vcard element of the preview text divided by 4
      top: calc(50% - (306px / 2) + (26px / 4));
      right: 4.5%;
    }
  }
}

@media screen and (max-height: 365px) and (max-width: 1000px) and (min-width: 751px) {
  main {
    .vcard {
      position: relative;
      top: 35px;
      right: auto;
    }
  }
}

@media screen and (max-width: 750px) {
  main {
    flex-direction: column;

    .hcard-builder {
      flex-basis: auto;
      order: 2;
    }

    .hcard-preview {
      flex-basis: auto;
      order: 1;
      padding: 40px 0 20px;
      min-width: auto;
    }
  }
}

@media screen and (max-width: 430px) {
  body {
    min-width: 320px;
  }

  main {
    @include single-row-form-fields();

    .hcard-builder {
      .form__section {
        padding-bottom: 24px;
      }

      .form__title {
        margin-bottom: 0;
      }
    }

    .vcard {
      width: 100%;
    }
  }
}
