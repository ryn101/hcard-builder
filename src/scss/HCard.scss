@import 'colours';
@import 'mixins';

$vcard-padding: 25px;
$photo-wrapper-right: 10px;
$photo-wrapper-width: 83px;

.vcard {
  width: 430px;
  background-color: white;
  border-bottom: 2px solid #9A9A9A;
  height: 306px;
  position: relative;

  .type {
    display: none;
  }

  .banner {
    position: absolute;
    top: -26px;
    right: 5px;
    color: #95A5A6;
  }

  .header {
    height: 100px;
    background-color: $sophisticated-navy;
    padding: 54px $vcard-padding 0 $vcard-padding;

    h1 {
      font-size: 18pt;
      margin: 0;
      color: #fff;
      font-family: Merriweather, Arial, serif;
      width: calc(100% - #{($photo-wrapper-right * 2) + $photo-wrapper-width - $vcard-padding});
      @include ellipsis();
    }
  }

  .photo-wrapper {
    border: 1px solid #9A9A9A;
    border-right-width: 2px;
    border-bottom-width: 2px;
    position: absolute;
    right: $photo-wrapper-right;
    top: 10px;
    height: 104px;
    width: $photo-wrapper-width;
    overflow: hidden;
    text-align: center;
    background-color: #e3e3e3;

    > img {
      height: 100%;
      width: 100%;
      object-fit: cover;
    }
  }

  .details {
    display: block;
    padding: $vcard-padding;
    margin: 0;
    font-style: normal;

    li {
      border-bottom: 1px $on-the-line-grey solid;
      padding-bottom: 1px;
      justify-content: left;

      &:not(:first-child) {
        margin-top: 12px;
      }
    }

    .detail {
      width: 100%;
      overflow: hidden;

      div > span {
        display: inline-block;
      }

      $detail-label-width: 76px;

      .label {
        width: $detail-label-width;

        color: $sophisticated-navy;
        text-transform: uppercase;
        font-size: 7.5pt;
      }

      .value {
        @include ellipsis();
        font-family: Merriweather, Arial, serif;
        width: calc(100% - #{$detail-label-width});
        vertical-align: text-bottom;
        margin-bottom: -5px;
        padding-bottom: 5px;
        min-height: 23px;
      }
    }
  }
}
