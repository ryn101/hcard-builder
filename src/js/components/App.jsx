import React, { Component } from 'react';
import '../../scss/App.scss';
import HCard from './HCard';
import HCardDetails from '../HCardDetails';
import HCardBuilderComponent from './HCardBuilder';

class App extends Component {
    constructor(props) {
        super(props);

        this.state = {
            hCard: new HCardDetails(),
            avatarFileUrl: undefined,
        };
    }

    /**
     * Event for handling the update of a hcard builder field
     */
    onFieldUpdate = (fieldId, fieldValue) => {
        const { hCard } = this.state;

        hCard[fieldId] = fieldValue;

        this.setState({ hCard });
    };

    /**
     * Event for handling the selection of an avatar via the file input
     *
     * @param fileUrl The file URL
     */
    onAvatarSelected = (fileUrl) => {
        const { avatarFileUrl } = this.state;

        if (avatarFileUrl) {
            URL.revokeObjectURL(avatarFileUrl);
        }

        this.setState({
            avatarFileUrl: URL.createObjectURL(fileUrl),
        });
    };

    onFormSubmission = (event) => event.preventDefault();

    render() {
        const { hCard, avatarFileUrl } = this.state;

        return (
            <>
                <HCardBuilderComponent
                    onFieldUpdate={this.onFieldUpdate}
                    onFormSubmission={this.onFormSubmission}
                    onAvatarSelected={this.onAvatarSelected}
                />
                <div className="hcard-preview">
                    <HCard hCard={hCard} avatarFileUrl={avatarFileUrl} />
                </div>
            </>
        );
    }
}

export default App;
