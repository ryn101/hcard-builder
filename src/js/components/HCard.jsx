import React from 'react';
import PropTypes from 'prop-types';
import HCardDetails from '../HCardDetails';

import '../../scss/HCard.scss';
import profilePicture from '../../images/profile-default.png';
import HCardProperty from './HCardProperty';
import HCardDetail from './HCardDetail';
import AddressState from '../AddressState';

function HCard(props) {
    const { hCard, avatarFileUrl } = props;
    const streetAddress = [hCard.houseNameOrNumber, hCard.street].filter(Boolean).join(' ');
    const fullName = [hCard.givenName, hCard.surname].filter(Boolean).join(' ');
    const state = new AddressState(hCard.state);

    return (
        <section className="vcard">
            <header className="vcard__header">
                <h2 className="vcard__heading fn" title={fullName}>{fullName}</h2>
                <div className="vcard__photo-wrapper">
                    <img alt="avatar" className="photo" src={avatarFileUrl} />
                </div>
            </header>
            <span className="vcard__banner">HCARD PREVIEW</span>
            <address className="adr">
                <ul className="vcard__details">
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Email" className="flex-row__column">
                            <HCardProperty name="email" value={hCard.email} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Phone" className="flex-row__column">
                            <HCardProperty name="tel">
                                <HCardProperty name="type" value="home" />
                                { hCard.phone }
                            </HCardProperty>
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Address" className="flex-row__column">
                            <HCardProperty name="street-address" value={streetAddress} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail className="flex-row__column">
                            <HCardProperty name="locality" value={hCard.suburb} />
                            <span>{hCard.suburb && hCard.state ? ', ' : ''}</span>
                            <HCardProperty name="region" value={state.name} abbreviation={state.abbreviation} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-2">
                        <HCardDetail label="Post Code" className="flex-row__column">
                            <HCardProperty name="postal-code" value={hCard.postcode} />
                        </HCardDetail>
                        <HCardDetail label="Country" className="flex-row__column">
                            <HCardProperty name="country" value={hCard.country} />
                        </HCardDetail>
                    </li>
                </ul>
            </address>
        </section>
    );
}

HCard.propTypes = {
    hCard: PropTypes.instanceOf(HCardDetails).isRequired,
    avatarFileUrl: PropTypes.string,
};

HCard.defaultProps = {
    avatarFileUrl: profilePicture,
};

export default HCard;
