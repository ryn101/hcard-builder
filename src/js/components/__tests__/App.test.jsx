import React from 'react';
import ReactDOM from 'react-dom';
import Adapter from 'enzyme-adapter-react-16';
import { configure, shallow } from 'enzyme';
import App from '../App';
import HCard from '../HCard';
import HCardBuilderComponent from '../HCardBuilder';

configure({ adapter: new Adapter() });

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
});

describe('once rendered', () => {
    let appNode;

    beforeEach(() => {
        appNode = shallow(
            <App />,
        );
    });

    it('renders with state correctly', () => {
        expect(appNode).toMatchSnapshot();
    });

    it('displays the builder and hcard', () => {
        expect(appNode.exists(HCardBuilderComponent)).toBeTruthy();
        expect(appNode.exists(HCard)).toBeTruthy();
    });
});
