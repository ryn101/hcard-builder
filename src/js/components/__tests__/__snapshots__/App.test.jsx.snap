// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once rendered renders with state correctly 1`] = `
<Fragment>
  <HCardBuilderComponent
    onAvatarSelected={[Function]}
    onFieldUpdate={[Function]}
    onFormSubmission={[Function]}
  />
  <div
    className="hcard-preview"
  >
    <HCard
      avatarFileUrl="test-file-stub"
      hCard={
        HCardDetails {
          "country": undefined,
          "email": undefined,
          "givenName": undefined,
          "houseNameOrNumber": undefined,
          "phone": undefined,
          "postcode": undefined,
          "state": undefined,
          "street": undefined,
          "suburb": undefined,
          "surname": undefined,
        }
      }
    />
  </div>
</Fragment>
`;
